File: rust/tools/golden/history.py
"""Golden outputs for weather history comparisons and anomaly callouts.

Writes rust/testdata/golden/history/*.json.
"""

from __future__ import annotations

import json
from datetime import date

from common import write

import accessiweather.weather_anomaly as weather_anomaly
from accessiweather.models.weather import CurrentConditions
from accessiweather.weather_history import (
    HistoricalWeatherData,
    WeatherComparison,
    WeatherHistoryService,
)


def comparisons():
    hist = HistoricalWeatherData(
        date=date(2026, 7, 14),
        temperature_max=80.0,
        temperature_min=60.0,
        temperature_mean=70.0,
        condition="Clear sky",
        humidity=None,
        wind_speed=10.0,
        wind_direction=180,
        pressure=None,
    )
    cases = [
        ("same", CurrentConditions(temperature=70.5, condition="Clear sky"), 1),
        ("warmer", CurrentConditions(temperature=75.25, condition="Clear sky"), 1),
        ("cooler_changed", CurrentConditions(temperature=62.0, condition="Rain"), 7),
        ("no_condition", CurrentConditions(temperature=69.0), 3),
        ("exactly_one_degree", CurrentConditions(temperature=71.0, condition="Fog"), 30),
    ]
    out = []
    for name, current, days_ago in cases:
        comparison = WeatherComparison.compare(current, hist, days_ago=days_ago)
        out.append(
            {
                "name": name,
                "current": current,
                "historical": hist,
                "days_ago": days_ago,
                "comparison": comparison,
                "summary": comparison.get_accessible_summary(),
            }
        )
    write("history", "comparisons", out)


class FakeArchiveClient:
    def __init__(self, response):
        self.response = response
        self.calls = []

    def _make_request(self, endpoint, params, use_archive=False):
        self.calls.append((endpoint, params, use_archive))
        if isinstance(self.response, Exception):
            raise self.response
        if callable(self.response):
            return self.response(params)
        return self.response

    @staticmethod
    def get_weather_description(code):
        return f"code {json.dumps(code)}"


def archive_parsing():
    target = date(2026, 7, 14)
    full = {
        "time": ["2026-07-14"],
        "weather_code": [3],
        "temperature_2m_max": [81.5],
        "temperature_2m_min": [62],
        "temperature_2m_mean": ["71.2"],
        "wind_speed_10m_max": [12.3],
        "wind_direction_10m_dominant": [185.7],
    }
    cases = [
        ("full", {"daily": full}),
        ("no_daily", {"hourly": {}}),
        ("empty_time", {"daily": {**full, "time": []}}),
        ("missing_mean", {"daily": {k: v for k, v in full.items() if k != "temperature_2m_mean"}}),
        ("null_code", {"daily": {**full, "weather_code": [None]}}),
        ("null_direction", {"daily": {**full, "wind_direction_10m_dominant": [None]}}),
        ("no_direction", {"daily": {k: v for k, v in full.items() if k != "wind_direction_10m_dominant"}}),
        ("bad_number", {"daily": {**full, "temperature_2m_max": ["hot"]}}),
        ("float_code", {"daily": {**full, "weather_code": [61.0]}}),
    ]
    out = []
    for name, response in cases:
        service = WeatherHistoryService(openmeteo_client=FakeArchiveClient(response))
        result = service.get_historical_weather(40.0, -75.0, target)
        out.append({"name": name, "date": target, "response": response, "result": result})
    write("history", "archive_parsing", out)


def anomaly():
    def responder(means_by_year):
        def respond(params):
            year = int(params["start_date"][:4])
            value = means_by_year.get(year)
            if value is None:
                return None
            return {"daily": {"temperature_2m_mean": value}}

        return respond

    cases = [
        ("five_years", 75.0, date(2026, 7, 15), date(2026, 7, 15), {
            2025: [70.0, 71.0, None], 2024: [72.0], 2023: [68.5, 69.5], 2022: [71.0], 2021: [70.0]}),
        ("insufficient", 75.0, date(2026, 7, 15), date(2026, 7, 15), {2025: [70.0], 2024: [72.0]}),
        ("empty_lists_skipped", 60.0, date(2026, 7, 15), date(2026, 7, 15), {
            2025: [], 2024: [None], 2023: [59.0], 2022: [60.2], 2021: [60.1]}),
        ("leap_day", 40.0, date(2028, 2, 29), date(2028, 2, 29), {
            2027: [45.0], 2026: [44.0], 2025: [46.0], 2024: [43.0], 2023: [47.0]}),
        ("recent_window_clamped", 50.0, date(2026, 7, 15), date(2025, 7, 20), {
            2025: [49.0], 2024: [48.0], 2023: [47.0], 2022: [51.0], 2021: [52.0]}),
    ]
    out = []
    for name, temp, current_date, today, means in cases:
        weather_anomaly.date = type(
            "FrozenDate", (date,), {"today": classmethod(lambda cls, t=today: t)}
        )
        client = FakeArchiveClient(responder(means))
        result = weather_anomaly.compute_anomaly(40.0, -75.0, temp, current_date, client)
        requests = [f"{p['start_date']}/{p['end_date']}" for _, p, _ in client.calls]
        responses = {}
        for _, params, _ in client.calls:
            responses[f"{params['start_date']}/{params['end_date']}"] = responder(means)(params)
        out.append(
            {
                "name": name,
                "current_temp_f": temp,
                "current_date": current_date,
                "today": today,
                "responses": responses,
                "requests": requests,
                "result": result,
            }
        )
    write("history", "anomaly", out)


if __name__ == "__main__":
    comparisons()
    archive_parsing()
    anomaly()

File: rust/crates/aw-providers/src/client/history.rs
//! Weather history comparisons and anomaly callouts against the Open-Meteo
//! archive. Port of `WeatherHistoryService` (`accessiweather/weather_history.py`)
//! and the fetch loop of `weather_anomaly.compute_anomaly`.
//!
//! Python constructs the service (when `weather_history_enabled`) but never
//! calls it, and nothing calls `compute_anomaly`; both are ported for
//! completeness.

use std::sync::Arc;

use aw_core::model::{AnomalyCallout, CurrentConditions, Location};
use aw_core::weather_anomaly;
use aw_core::weather_history::{
    parse_historical_weather, HistoricalWeatherData, WeatherComparison, HISTORY_DAILY_FIELDS,
};
use chrono::{Duration, NaiveDate};
use serde_json::Value;

use super::sources::SourceResult;

/// One archive request (`GET {archive}/archive`, `timezone=auto`).
#[derive(Debug, Clone, PartialEq)]
pub struct ArchiveRequest {
    pub latitude: f64,
    pub longitude: f64,
    pub start_date: NaiveDate,
    pub end_date: NaiveDate,
    /// Daily variables, sent as repeated `daily=` parameters.
    pub daily: Vec<&'static str>,
    /// "fahrenheit" or "celsius".
    pub temperature_unit: String,
}

/// Open-Meteo archive API (`OpenMeteoApiClient._make_request("archive", ...,
/// use_archive=True)`; Python uses a 30 s timeout for history).
pub trait ArchiveSource: Send + Sync {
    /// The response JSON.
    fn archive(&self, request: &ArchiveRequest) -> SourceResult<Value>;
    /// `OpenMeteoApiClient.get_weather_description` for a WMO code value.
    fn weather_description(&self, code: &Value) -> String;
}

pub struct WeatherHistoryService {
    archive: Arc<dyn ArchiveSource>,
}

impl WeatherHistoryService {
    pub fn new(archive: Arc<dyn ArchiveSource>) -> Self {
        Self { archive }
    }

    /// `get_historical_weather`: one day from the archive, `None` when the
    /// request fails or required values are missing.
    pub fn get_historical_weather(
        &self,
        latitude: f64,
        longitude: f64,
        target_date: NaiveDate,
        temperature_unit: &str,
    ) -> Option<HistoricalWeatherData> {
        let request = ArchiveRequest {
            latitude,
            longitude,
            start_date: target_date,
            end_date: target_date,
            daily: HISTORY_DAILY_FIELDS.to_vec(),
            temperature_unit: temperature_unit.to_string(),
        };
        let response = match self.archive.archive(&request) {
            Ok(response) => response,
            Err(e) => {
                tracing::error!("Failed to fetch historical weather data: {e}");
                return None;
            }
        };
        let parsed = parse_historical_weather(&response, target_date, |code| {
            self.archive.weather_description(code)
        });
        if parsed.is_none() {
            tracing::warn!("No historical data available for {target_date}");
        }
        parsed
    }

    /// Compare with `today - 1 day` (`compare_with_yesterday`).
    pub fn compare_with_yesterday(
        &self,
        location: &Location,
        current: &CurrentConditions,
        temperature_unit: &str,
        today: NaiveDate,
    ) -> Option<WeatherComparison> {
        self.compare_days_ago(location, current, temperature_unit, today, 1)
    }

    /// Compare with `today - 7 days` (`compare_with_last_week`).
    pub fn compare_with_last_week(
        &self,
        location: &Location,
        current: &CurrentConditions,
        temperature_unit: &str,
        today: NaiveDate,
    ) -> Option<WeatherComparison> {
        self.compare_days_ago(location, current, temperature_unit, today, 7)
    }

    /// Compare with a specific date (`compare_with_date`).
    pub fn compare_with_date(
        &self,
        location: &Location,
        current: &CurrentConditions,
        target_date: NaiveDate,
        temperature_unit: &str,
        today: NaiveDate,
    ) -> Option<WeatherComparison> {
        let days_ago = (today - target_date).num_days();
        let historical = self.get_historical_weather(
            location.latitude,
            location.longitude,
            target_date,
            temperature_unit,
        )?;
        WeatherComparison::compare(current, &historical, days_ago)
    }

    fn compare_days_ago(
        &self,
        location: &Location,
        current: &CurrentConditions,
        temperature_unit: &str,
        today: NaiveDate,
        days: i64,
    ) -> Option<WeatherComparison> {
        let date = today - Duration::days(days);
        let historical = self.get_historical_weather(
            location.latitude,
            location.longitude,
            date,
            temperature_unit,
        )?;
        WeatherComparison::compare(current, &historical, days)
    }
}

/// `compute_anomaly`: current temperature versus the five-year archive
/// baseline for the same ±7 days.
pub fn compute_anomaly(
    archive: &dyn ArchiveSource,
    latitude: f64,
    longitude: f64,
    current_temp_f: f64,
    current_date: NaiveDate,
    today: NaiveDate,
) -> Option<AnomalyCallout> {
    weather_anomaly::compute_anomaly(current_temp_f, current_date, today, |start, end| {
        let request = ArchiveRequest {
            latitude,
            longitude,
            start_date: start,
            end_date: end,
            daily: vec!["temperature_2m_mean"],
            temperature_unit: "fahrenheit".into(),
        };
        archive
            .archive(&request)
            .map_err(|e| tracing::debug!("Anomaly archive fetch failed: {e}"))
            .ok()
    })
}

#[cfg(test)]
mod tests {
    use std::sync::Mutex;

    use aw_core::golden::{self, field};

    use super::*;
    use crate::client::SourceError;

    struct FakeArchive {
        response: SourceResult<Value>,
        requests: Mutex<Vec<ArchiveRequest>>,
    }

    impl ArchiveSource for FakeArchive {
        fn archive(&self, request: &ArchiveRequest) -> SourceResult<Value> {
            self.requests.lock().unwrap().push(request.clone());
            self.response.clone()
        }
        fn weather_description(&self, code: &Value) -> String {
            format!("code {code}")
        }
    }

    fn service(response: SourceResult<Value>) -> (Arc<FakeArchive>, WeatherHistoryService) {
        let fake = Arc::new(FakeArchive {
            response,
            requests: Mutex::default(),
        });
        (fake.clone(), WeatherHistoryService::new(fake))
    }

    #[test]
    fn golden_archive_responses_through_the_service() {
        for case in golden::load("history/archive_parsing.json")
            .as_array()
            .unwrap()
        {
            let (fake, service) = service(Ok(case["response"].clone()));
            let date: NaiveDate = field(case, "date");
            let result = service.get_historical_weather(40.0, -75.0, date, "fahrenheit");
            let expected: Option<HistoricalWeatherData> = field(case, "result");
            assert_eq!(result, expected, "{}", case["name"]);
            let request = &fake.requests.lock().unwrap()[0];
            assert_eq!(request.daily, HISTORY_DAILY_FIELDS.to_vec());
            assert_eq!((request.start_date, request.end_date), (date, date));
        }
    }

    #[test]
    fn failed_request_gives_none_and_comparisons_use_today() {
        let (_, failing) = service(Err(SourceError::new("offline")));
        let today = NaiveDate::from_ymd_opt(2026, 7, 15).unwrap();
        assert!(failing
            .get_historical_weather(1.0, 2.0, today, "celsius")
            .is_none());

        let response = serde_json::json!({"daily": {
            "time": ["2026-07-08"], "weather_code": [0], "temperature_2m_max": [80.0],
            "temperature_2m_min": [60.0], "temperature_2m_mean": [70.0], "wind_speed_10m_max": [5.0]}});
        let (fake, service) = service(Ok(response));
        let location = Location::new("Here", 40.0, -75.0);
        let current = CurrentConditions {
            temperature: Some(75.0),
            condition: Some("code 0".into()),
            ..Default::default()
        };
        let week = service
            .compare_with_last_week(&location, &current, "fahrenheit", today)
            .unwrap();
        assert_eq!(
            week.accessible_summary(),
            "Compared to last week: 5.0 degrees warmer."
        );
        assert_eq!(
            fake.requests.lock().unwrap()[0].start_date,
            NaiveDate::from_ymd_opt(2026, 7, 8).unwrap()
        );
        let dated = service
            .compare_with_date(
                &location,
                &current,
                NaiveDate::from_ymd_opt(2026, 7, 5).unwrap(),
                "fahrenheit",
                today,
            )
            .unwrap();
        assert_eq!(dated.days_ago, 10);
    }
}
